src/what-unsafe-does.md

```diff
@@ -16,19 +16,30 @@ to your program. You definitely *should not* invoke Undefined Behavior.
 Unlike C, Undefined Behavior is pretty limited in scope in Rust. All the core
 language cares about is preventing the following things:
 
-* Dereferencing null, dangling, or unaligned pointers
+* Dereferencing (using the `*` operator on) null, dangling, or unaligned
+  pointers
 * Reading [uninitialized memory][]
 * Breaking the [pointer aliasing rules][]
-* Producing invalid primitive values:
-    * dangling/null references
+* Producing invalid primitive values (either alone or as a field of a compound
+  type such as `enum`/`struct`/array/tuple):
+    * dangling/null/unaligned references
     * null `fn` pointers
     * a `bool` that isn't 0 or 1
     * an undefined `enum` discriminant
     * a `char` outside the ranges [0x0, 0xD7FF] and [0xE000, 0x10FFFF]
-    * A non-utf8 `str`
+    * a non-utf8 `str`
+    * an invalid library type with custom invalid values, such as a `NonNull` or
+      `NonZero*` that is 0
 * Unwinding into another language
 * Causing a [data race][race]
```

Comment:
Calling a `#[target_feature(enable = "foo")] unsafe fn bar();` function on a thread of execution that does not support the target-feature `foo` is undefined behavior as well. 

It's not clear to me whether the `target_feature` attribute makes this a precondition on calling the function - all these functions are `unsafe fn` anyways - or whether this is something worth documenting here.

Reply:
Good point; this is listed in the other UB list at https://doc.rust-lang.org/reference/behavior-considered-undefined.html. @Gankro what do you think?